File: javascript/design/170. 两数之和 III - 数据结构设计.js
/**
 * Initialize your data structure here.
 */
var TwoSum = function() {
    this.res = []
    this.isSorted = false
};

/**
 * Add the number to an internal data structure..
 * @param {number} number
 * @return {void}
 */
TwoSum.prototype.add = function(number) {
    this.res.push(number)
    this.isSorted = false
};

/**
 * Find if there exists any pair of numbers which sum is equal to the value.
 * @param {number} value
 * @return {boolean}
 */
TwoSum.prototype.find = function(value) {
    if(this.isSorted === false){
        this.res.sort(function (a,b){
            return a-b
        })
        this.isSorted = true
    }
    let i = 0
    let j = this.res.length -1
    while(i<j){
        let t = this.res[i] + this.res[j]
        if(t===value){
            return true
        }else if(t > value){
            j--
        }else{
            i++
        }
    }
    return false
};

/**
 * Your TwoSum object will be instantiated and called as such:
 * var obj = new TwoSum()
 * obj.add(number)
 * var param_2 = obj.find(value)
 */

File: go/design/170. 两数之和 III - 数据结构设计.go
package design

import "sort"

type TwoSum struct {
	res      []int
	isSorted bool
}

/** Initialize your data structure here. */
func Constructor() TwoSum {
	return TwoSum{
		res:      make([]int, 0),
		isSorted: false,
	}
}

/** Add the number to an internal data structure.. */
func (this *TwoSum) Add(number int) {
	this.res = append(this.res, number)
	this.isSorted = false
}

/** Find if there exists any pair of numbers which sum is equal to the value. */
func (this *TwoSum) Find(value int) bool {
	if this.isSorted == false {
		sort.Ints(this.res)
		this.isSorted = true
	}
	i, j := 0, len(this.res)-1
	for i < j {
		t := this.res[i] + this.res[j]
		if t == value {
			return true
		} else if t > value {
			j--
		} else {
			i++
		}
	}
	return false
}
